README.md
# ssc-formula2 <sup>[![Version Badge][npm-version-svg]][package-url]</sup>

[![Build Status][travis-svg]][travis-url]
[![dependency status][deps-svg]][deps-url]
[![dev dependency status][dev-deps-svg]][dev-deps-url]
[![License][license-image]][license-url]
[![Downloads][downloads-image]][downloads-url]

[![npm badge][npm-badge-png]][package-url]

> An formula editor for the Yonyou.

![ssc-formula2 in action](https://raw.githubusercontent.com/yyssc/ssc-formula2/master/ssc-formula2-demo.gif)

## Live Playground

For examples of the formula in action, go to https://yyssc.github.io/ssc-formula2.

OR

To run that demo on your own computer:
* Clone this repository
* `npm install`
* `npm run storybook`
* Visit http://localhost:9001/

## Getting Started
### Install dependencies
Ensure packages are installed with correct version numbers by running:
  ```sh
  (
    export PKG=ssc-formula2;
    npm info "$PKG" peerDependencies --json | command sed 's/[\{\},]//g ; s/: /@/g; s/ *//g' | xargs npm install --save "$PKG"
  )
  ```

  Which produces and runs a command like:

  ```sh
  npm install --save ssc-formula2 moment@>=#.## react@>=#.## react-dom@>=#.## ssc-refer2@>=#.## rc-tree@>=#.##
  ```

### Include component
```js
import { Formula } from 'ssc-formula2';
```

### Include styles
```js
import 'rc-tree/assets/index.css';
```

## 组件结构

- Formula
  - TextBox
  - FormulaTabs
    - 单据字段Tab
      - SearchBox
      - FormulaTree
    - 固定值Tab
      - [Refers]
    - 档案转换Tab

[package-url]: https://npmjs.org/package/ssc-formula2
[npm-version-svg]: http://versionbadg.es/yyssc/ssc-formula2.svg
[travis-svg]: https://travis-ci.org/yyssc/ssc-formula2.svg
[travis-url]: https://travis-ci.org/yyssc/ssc-formula2
[deps-svg]: https://david-dm.org/yyssc/ssc-formula2.svg
[deps-url]: https://david-dm.org/yyssc/ssc-formula2
[dev-deps-svg]: https://david-dm.org/yyssc/ssc-formula2/dev-status.svg
[dev-deps-url]: https://david-dm.org/yyssc/ssc-formula2#info=devDependencies
[npm-badge-png]: https://nodei.co/npm/ssc-formula2.png?downloads=true&stars=true
[license-image]: http://img.shields.io/npm/l/ssc-formula2.svg
[license-url]: LICENSE
[downloads-image]: http://img.shields.io/npm/dm/ssc-formula2.svg
[downloads-url]: http://npm-stat.com/charts.html?package=ssc-formula2


CHANGELOG.md
# Change Log

## v0.1.20

- [fix] failed to npm publish ([#23](https://github.com/yyssc/ssc-formula2/issues/23))

## v0.1.19

- [chore] reduce bundle size ([#22](https://github.com/yyssc/ssc-formula2/issues/22))

## v0.1.18

- [new] custom input field and overlay style ([#18](https://github.com/yyssc/ssc-formula2/issues/18))
- [fix] FICLOUD-1019 档案转换插入时候应该是node的path而不是code ([#19](https://github.com/yyssc/ssc-formula2/issues/19))

## v0.1.17

- [new] 使用`DropdownTree`组件替换`ReferList`组件 ([#17](https://github.com/yyssc/ssc-formula2/issues/17))

## v0.1.16

- [fix] 修复getID函数第一个参数是null的问题 ([#16](https://github.com/yyssc/ssc-formula2/issues/16))

## v0.1.15

- [fix] 修复布局问题 ([#15](https://github.com/yyssc/ssc-formula2/issues/15))
- [fix] classtype取得是name不是code ([#15](https://github.com/yyssc/ssc-formula2/issues/15))

## v0.1.14

- [new] 在“档案转换”标签页添加参照 ([#15](https://github.com/yyssc/ssc-formula2/issues/15))

## v0.1.13

- [new] 从yzb-comp引用业务层的参照组件

## v0.1.12

- [fix] FICLOUD-1019 测试档案转换插入值的格式 test01  ([#15](https://github.com/yyssc/ssc-formula2/issues/15))

## v0.1.11

- [fix] FilterAccSubjectRefer not found  ([#14](https://github.com/yyssc/ssc-formula2/issues/14))

## v0.1.10

- [new] 添加科目表作为科目的过滤条件 ([#13](https://github.com/yyssc/ssc-formula2/issues/13))

## v0.1.9

- [new] 档案转换最后一个字段不需要双引号 ([#12](https://github.com/yyssc/ssc-formula2/issues/12))

## v0.1.8

- [new] 拼装节点所在的路径 ([#11](https://github.com/yyssc/ssc-formula2/issues/11))

## v0.1.7

- [new] impl cancel and validate button ([#10](https://github.com/yyssc/ssc-formula2/issues/10))

## v0.1.6

- [fix] 解决固定值Tab中档案值参照出现undefined问题 ([#8](https://github.com/yyssc/ssc-formula2/issues/8))
- [new] 档案转换tab页中选中一个item时候显示详细信息 ([#9](https://github.com/yyssc/ssc-formula2/issues/9))

## v0.1.5

- [fix] Add prop 固定值档案值refCode for Formula component

## v0.1.4

- [new] 临时方案，由外面传进来档案值refCode

## v0.1.3

- [fix] Fix building error in UglifyJs ([#7](https://github.com/yyssc/ssc-formula2/issues/7))

## v0.1.2

- [new] Update TextBox value according to "单据字段" or "档案转换" selected items
- [fix] make textbox auto height

## v0.1.1
- [fix] Remove peer denpendency `react-addons-shallow-compare` ([#2](https://github.com/yyssc/ssc-formula2/issues/2))
- [new] Export `档案转换ItemsShape` ([#3](https://github.com/yyssc/ssc-formula2/issues/3))
- [new] Use `固定值ReferDataUrl` prop to set Refer URL ([#4](https://github.com/yyssc/ssc-formula2/issues/4))
- [fix] Put `rc-tree` in peer dependency ([#5](https://github.com/yyssc/ssc-formula2/issues/5))

## v0.1.0
 - Initial commit


stories/DayPickerRangeController.js
import React from 'react';
import moment from 'moment';
import { storiesOf, action } from '@kadira/storybook';

import isSameDay from '../src/utils/isSameDay';
import isInclusivelyAfterDay from '../src/utils/isInclusivelyAfterDay';

import { VERTICAL_ORIENTATION } from '../constants';

import DayPickerRangeControllerWrapper from '../examples/DayPickerRangeControllerWrapper';

const monospace = text => `<span style="font-family:monospace;background:#f7f7f7">${text}</span>`;

const dayPickerRangeControllerInfo = `The ${monospace('DayPickerRangeController')} component is a
  fully controlled version of the ${monospace('DayPicker')} that has built-in rules for selecting a
  date range. Unlike the ${monospace('DayPicker')}, which requires the consumer to explicitly define
  ${monospace('onDayMouseEnter')}, ${monospace('onDayMouseLeave')}, and ${monospace('onDayClick')}
  handlers, the consumer needs simply to maintain the ${monospace('focusedInput')},
  ${monospace('startDate')}, and ${monospace('endDate')} values in state and then pass these down as
  props along with ${monospace('onFocusChange')} and ${monospace('onDatesChange')} callbacks that
  update them appropriately. You can see an example of this implementation <a href=
  "https://github.com/yyssc/ssc-formula2/blob/master/examples/DayPickerRangeControllerWrapper.jsx">
  here</a>. <br/><br/>
  Note that the ${monospace('focusedInput')} prop may be ${monospace('null')}, but if this is the
  case, dates are not selectable. As a result, in the example wrapper, we always force
  ${monospace('focusedInput')} to be truthy in the ${monospace('onFocusChange')} method. <br/><br/>
  The ${monospace('DayPickerRangeController')} is particularly useful if you are interested in the
  ${monospace('DateRangePicker')} functionality and calendar presentation, but would like to
  implement your own inputs.`;

const InfoPanel = ({ text }) => (
  <div
    style={{
      backgroundColor: '#fff',
      fontColor: '#3c3f40',
      fontSize: 14,
      margin: '8px 0',
      padding: 16,
    }}
  >
    <span dangerouslySetInnerHTML={{ __html: text }} />
  </div>
);

const InfoPanelDecorator = story => (
  <div>
    <InfoPanel text={dayPickerRangeControllerInfo} />
    {story()}
  </div>
);

const TestPrevIcon = () => (
  <span
    style={{
      border: '1px solid #dce0e0',
      backgroundColor: '#fff',
      color: '#484848',
      padding: '3px',
    }}
  >
    Prev
  </span>
);

const TestNextIcon = () => (
  <span
    style={{
      border: '1px solid #dce0e0',
      backgroundColor: '#fff',
      color: '#484848',
      padding: '3px',
    }}
  >
    Next
  </span>
);

const TestCustomInfoPanel = () => (
  <div
    style={{
      padding: '10px 21px',
      borderTop: '1px solid #dce0e0',
      color: '#484848',
    }}
  >
    &#x2755; Some useful info here
  </div>
);

const datesList = [
  moment(),
  moment().add(1, 'days'),
  moment().add(3, 'days'),
  moment().add(9, 'days'),
  moment().add(10, 'days'),
  moment().add(11, 'days'),
  moment().add(12, 'days'),
  moment().add(13, 'days'),
];

storiesOf('DayPickerRangeController', module)
  .addDecorator(InfoPanelDecorator)
  .addWithInfo('default', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
    />
  ))
  .addWithInfo('with custom inputs', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      showInputs
    />
  ))
  .addWithInfo('non-english locale', () => {
    moment.locale('zh-cn');
    return (
      <DayPickerRangeControllerWrapper
        onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
        onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
        onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
        monthFormat="YYYY[年]MMMM"
      />
    );
  })
  .addWithInfo('single month', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      numberOfMonths={1}
    />
  ))
  .addWithInfo('3 months', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      numberOfMonths={3}
    />
  ))
  .addWithInfo('vertical', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      orientation={VERTICAL_ORIENTATION}
    />
  ))
  .addWithInfo('with custom month navigation', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      navPrev={<TestPrevIcon />}
      navNext={<TestNextIcon />}
    />
  ))
  .addWithInfo('with outside days enabled', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      numberOfMonths={1}
      enableOutsideDays
    />
  ))
  .addWithInfo('with month specified on open', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      initialVisibleMonth={() => moment('04 2017', 'MM YYYY')}
    />
  ))
  .addWithInfo('with minimum nights set', () => (
    <DayPickerRangeControllerWrapper
      minimumNights={3}
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      initialStartDate={moment().add(3, 'days')}
      autoFocusEndDate
    />
  ))
  .addWithInfo('allows single day range', () => (
    <DayPickerRangeControllerWrapper
      minimumNights={0}
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      initialStartDate={moment().add(3, 'days')}
      autoFocusEndDate
    />
  ))
  .addWithInfo('allows all days, including past days', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      isOutsideRange={() => false}
    />
  ))
  .addWithInfo('allows next two weeks only', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      isOutsideRange={day =>
        !isInclusivelyAfterDay(day, moment()) ||
        isInclusivelyAfterDay(day, moment().add(2, 'weeks'))
      }
    />
  ))
  .addWithInfo('with some blocked dates', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      isDayBlocked={day1 => datesList.some(day2 => isSameDay(day1, day2))}
    />
  ))
  .addWithInfo('with some highlighted dates', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      isDayHighlighted={day1 => datesList.some(day2 => isSameDay(day1, day2))}
    />
  ))
  .addWithInfo('blocks fridays', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      isDayBlocked={day => moment.weekdays(day.weekday()) === 'Friday'}
    />
  ))
  .addWithInfo('with custom daily details', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      renderDay={day => day.format('ddd')}
    />
  ))
  .addWithInfo('with info panel', () => (
    <DayPickerRangeControllerWrapper
      onOutsideClick={action('DayPickerRangeController::onOutsideClick')}
      onPrevMonthClick={action('DayPickerRangeController::onPrevMonthClick')}
      onNextMonthClick={action('DayPickerRangeController::onNextMonthClick')}
      renderCalendarInfo={() => (
        <TestCustomInfoPanel />
      )}
    />
  ));
